### ISSUE TITLE
fix socket timeout exceptions on failed stream close on kubernetes

### ISSUE THREAD

**Issue by jrhizor** [airbytehq/airbyte#11621]
The underlying issue for https://github.com/airbytehq/oncall/issues/140 is possible to reproduce across a portion of syncs run in load testing.

**Comment by jrhizor** [comment]
Also confirmed we are seeing this occasionally on Cloud still (according to worker logs). Instances of the failure can be found in GCP log explorer with:
```
resource.type="k8s_container"
resource.labels.project_id="prod-ab-cloud-proj"
resource.labels.location="us-west3"
resource.labels.cluster_name="prod-gke-cluster"
resource.labels.namespace_name="ab"
"java.net.SocketException: Connection timed out"
```

**Comment by jrhizor** [comment]
see https://github.com/airbytehq/airbyte/pull/11083#issuecomment-1084422213 for more info

**Comment by jrhizor** [comment]
For current sprint: 
- Try to come up with an approach to debug the issue / see how we can instrument these syncs to see what's happening better. 
- Also it may be worthwhile to figure out how we can get this down to a level where we can reproduce this more cheaply without upgrading nodes massively and running hundreds of syncs over and over.
- I likely won't have time to fix this and may need to set this up to hand off to someone to dive in more depth and fix.

**Comment by jrhizor** [comment]
Due to the [analysis for the first hypothesis](https://docs.google.com/spreadsheets/d/1CuFe3yZQoUUSv5oSeSsdehys2BSL5rGTY3U167w79io/edit#gid=0) @git-phu is going to help with inspecting the pods and the nodes to see if there are any suspicious things happening on the `dev-2` cluster vs the `dev` cluster. 

If that doesn't work, we'll roll all of the nodes in the `dev-2` cluster and see if there are node-level issues causing this problem.

If these investigations today don't lead to anything then I'll need to hand this off and go back to the brainstorming phase.

**Comment by jrhizor** [comment]
For runs that don't have socket exception errors, we aren't remotely close to limits. From the last running entry for `watch "kubectl top -n jobs pod source-postgres-strict-encrypt-sync-19053-0-azorh --containers"`:
```
POD                                                 NAME                    CPU(cores)   MEMORY(bytes)

source-postgres-strict-encrypt-sync-19053-0-azorh   call-heartbeat-server   6m           1Mi

source-postgres-strict-encrypt-sync-19053-0-azorh   main                    145m         105Mi

source-postgres-strict-encrypt-sync-19053-0-azorh   relay-stderr            0m           1Mi

source-postgres-strict-encrypt-sync-19053-0-azorh   relay-stdout            1m           1Mi
```

Where our default limits are:
```
  private static final String DEFAULT_MEMORY_REQUEST = "25Mi";
  private static final String DEFAULT_MEMORY_LIMIT = "50Mi";
  private static final String DEFAULT_CPU_REQUEST = "0.1";
  private static final String DEFAULT_CPU_LIMIT = "0.2";
```

**Comment by jrhizor** [comment]
For small jobs, I ran with "empty" streams from the source database. Two 10 run batches and then two 15 run batches. They complete in ~2min. I ran these on `dev-2`. 

None of them ran into the issue. 

Then just to double check, I re-ran the 10 run batch with the 600k row table (~7min to the point of getting stuck) and saw 4 failures again.

It may be worth creating a separate table and search for the point where row count (or at least run length) might impact the possibility of success).

Good call to try this out @supertopher

**Comment by jrhizor** [comment]
Here are some remaining areas to explore:
- Find at which table size this problem starts to happen regularly 
- Retry testing with non-Postgres Java sources
- Retry testing with non-Postgres Python sources
- Add an environment variable toggle for the `keepalive` so we can keep the failing version open longer
- Try to compare version information for tests run on `dev` vs `dev-2` (everything from the underlying GKE nodes to the Airbyte version running)
- Enable verbose logging in socat sidecars?
- The biggest question is what we can do at the replication process level to log/understand exactly what is happening with the packets at the conclusion of the sync on both pods. Ideally we should be able to compare exactly what each side says happened and see the discrepancy. 

I added this to both of the boards in a "to do" capacity. 

@benmoriceau and @pmossman: I'll leave it to you two to figure out who should work on this. @git-phu and @supertopher have some context (as well as @benmoriceau). We should have one person to lead this investigation and be responsible for making sure this debugging is efficient and results in finally closing this issue and the root cause of https://github.com/airbytehq/oncall/issues/140. In my opinion, this issue is high priority since we still don't know what's causing it so it could come back to bite us at a much higher rate.

**Comment by git-phu** [comment]
New updates:
Did some more testing in `dev-2`.
`dev-2` environment changes:
- [preserved logs for source pods](https://github.com/airbytehq/airbyte/compare/peter/tee?expand=1) using [tee](https://linux.die.net/man/1/tee)
- created new node pool (using `e2-standard-4` nodes) for supporting greater load during testing

Tests using the postgres source connector:
- Using the current version of the postgres connector (0.1.7) versus the latest version (0.1.12) didn't seem to affect the failure rate (around 30%)
- comparing the k8s pod logs of the source pod to the orchestrator pod's view of the source pod logs, it does look like whenever we have this socket failure, the orchestrator is missing some logs that are present in the k8s pod logs.  They are typically the last few lines of the log:
    ```
    2022-04-27 00:28:02 INFO i.a.i.s.r.AbstractDbSource(lambda$createReadIterator$7):250 - Reading stream fifty_thousand_row_table. Records read: 600000
    2022-04-27 00:28:02 INFO i.a.i.s.r.AbstractDbSource(lambda$read$2):124 - Closing database connection pool.
    2022-04-27 00:28:02 INFO i.a.i.s.r.AbstractDbSource(lambda$read$2):126 - Closed database connection pool.
    2022-04-27 00:28:02 INFO i.a.i.b.IntegrationRunner(runInternal):169 - Completed integration: io.airbyte.integrations.source.postgres.PostgresSourceStrictEncrypt
    2022-04-27 00:28:02 INFO i.a.i.s.p.PostgresSourceStrictEncrypt(main):35 - completed source: class io.airbyte.integrations.source.postgres.PostgresSourceStrictEncrypt
    ```
    - There were a few rare cases where these last few lines were also missing from the k8s pod logs
- Based on this, one hypothesis was that the Java process in the source pod might be exiting too quickly (before we can capture all the output in the orchestrator), so I tested out some modifications to the connector.
    - When adding a 1 minute sleep to the end of the main method in the PostgresSourceStrictEncrypt connector, all the syncs complete successfully (20/20).
    - When decreasing the sleep to 1 second, failures crop up again, but at a much lower rate (1/21).
    - So it does seem that having a sleep at the end of the main method somehow affects the failure rate

Testing other source connectors:
- Currently testing redshift source.  (No failures so far but don't have enough runs yet to conclude anything).

**Comment by git-phu** [comment]
Some followup/related PRs:

- https://github.com/airbytehq/airbyte/pull/12632
    - Main PR: Implementing the change for the orchestrator to close the network socket

Nice to haves:
- https://github.com/airbytehq/airbyte/pull/12638
    - socat logging
- https://github.com/airbytehq/airbyte/pull/12478
    - source pod logging
- https://github.com/airbytehq/airbyte/pull/11262
    - this is how I was able to build images locally for faster iteration

Other potential followups:
- increase the minimum cpu we give to orchestrator pods
   - in testing this particular postgres source (an rds instance with ~40MB of test data), I observed that:
       - with 200m cpu, syncs took around 7-10 minutes on avg
       - with 500m or more cpu, syncs took around 2.5 minutes avg
    - might be worth looking into bumping the min cpu we give to sync
- do not set a cpu limit for orchestrator pods
    - from the links in my previous comment, it seems like the current consensus in the k8s community is that for pods that should have guaranteed resources, we should not set cpu limits
    - with the main PR fix we shouldn't require this to address the socket timeout issue anymore, but maybe it's a good idea to do anyway for more consistent performance with syncs

**Comment by cgardens** [comment]
@git-phu Before your change we already called `Source.close()` later in the code (source is in a try-with-resource [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-workers/src/main/java/io/airbyte/workers/general/DefaultReplicationWorker.java#L141)). Is the key that calling `Source.close()` early was necessary or is it just that the greater time out saves us?

**Comment by git-phu** [comment]
@cgardens Yes you are right.  The important part is the longer timeout on socat (so that the source doesn't end the connection after 500ms) which gives the orchestrator enough time to receive all messages and be the one to initiate the socket close.  

Explicitly calling the `Source.close()` method earlier isn't strictly necessary.  I just felt doing so would make it easier to verify that everything is working as expected when the orchestrator closes the socket at the time the orchestrator reads the EOF signal from the source.  This way we can check the socat logs on the source and verify that the orchestrator closes the connection at around the same time the source sends the EOF. 

In practice though things should still work if we just leave the socket to the source open until the try-with-resource block closes it, which I believe only closes the source after both the source and destination complete (please let me know if I'm mistaken).  This could lead to socat on the source sometimes still timing out, which shouldn't be a problem since the configured timeout is now much higher than 500ms, but I felt it could be confusing to have a setup where sometimes (but not always) reaching the timeout is not indicative of a problem.
